# Problem 1 - Word Counts

## Part A. Characters in Little Women

In [1]:
!wget https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/women.txt

--2016-09-23 13:12:34--  https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/women.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.20.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.20.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1053440 (1.0M) [text/plain]
Saving to: ‘women.txt’

women.txt           100%[=====================>]   1.00M  --.-KB/s   in 0.05s  

2016-09-23 13:12:35 (21.9 MB/s) - ‘women.txt’ saved [1053440/1053440]



In [15]:
!grep -o "Jo " women.txt | wc 

    652     652    2608


In [16]:
!grep -o "Beth " women.txt | wc

    227     227    1362


In [17]:
!grep -o "Meg " women.txt | wc

    351     351    1755


In [18]:
!grep -o "Amy " women.txt | wc

    333     333    1665


Within the text Little Women, Jo is mentioned 652 times, Beth is mentioned 227 times, Meg is mentioned 351 times, and Amy is mentioned 333 times.

I used the "grep -o" command which counts each time the word shows up in each line. I did not use the command that is given within the project file because it kept failing on me and not outputting the right information. I found that I got what I think is the correct answer by just utilizing the first part of the command and then accounting for white space after the words to ensure it captures only those words. I do not account for white space before the words because if they are at the beginning of a paragraph I do not believe that there would be white space before it in the line.

At first Jo printed out over 1000 times and I was not sure why it was so high. I then realized that I didnt account for the fact that it would also count words that began with Jo, so I had to encapsulate the search with spaces before and after. 

## Part B. Juliet and Romeo in Romeo and Julie

In [2]:
!wget https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/romeo.txt

--2016-09-23 13:12:37--  https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/romeo.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.20.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.20.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 178983 (175K) [text/plain]
Saving to: ‘romeo.txt’

romeo.txt           100%[=====================>] 174.79K  --.-KB/s   in 0.02s  

2016-09-23 13:12:37 (8.07 MB/s) - ‘romeo.txt’ saved [178983/178983]



In [7]:
!grep Rom. romeo.txt | wc

    312    2408   14353


In [8]:
!grep Jul. romeo.txt | wc

    189    1511    8915


Because this is a play and we want to know the amount of times Romeo and Juliet have speaking lines, we have to see how the play script denotes when they each speak. In this text, they do so using the first 3 letters of their name and a period, so instead of seeing where their names are in the text, we get a word count for Rom. and Jul..

# Problem 2 - Capital Bikeshare

In [3]:
!wget https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/2016q1.csv.zip

--2016-09-23 13:12:41--  https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/2016q1.csv.zip
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.20.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.20.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 10643003 (10M) [application/octet-stream]
Saving to: ‘2016q1.csv.zip’

2016q1.csv.zip      100%[=====================>]  10.15M  65.5MB/s   in 0.2s   

2016-09-23 13:12:41 (65.5 MB/s) - ‘2016q1.csv.zip’ saved [10643003/10643003]



## Part A

In [3]:
!unzip 2016q1.csv.zip

Archive:  2016q1.csv.zip
  inflating: 2016q1.csv              


In [3]:
!csvcut -c5 2016q1.csv | sort | uniq -c | sort -n | tail -10 

   5514 17th & Corcoran St NW
   5649 Eastern Market Metro / Pennsylvania Ave & 7th St SE
   6491 New Hampshire Ave & T St NW
   6568 14th & V St NW
   7401 15th & P St NW
   7479 Thomas Circle
   8138 Jefferson Dr & 14th St SW
   9388 Lincoln Memorial
   9560 Massachusetts Ave & Dupont Circle NW
  13120 Columbus Circle / Union Station


In [15]:
!csvcut -c7 2016q1.csv | sort | uniq -c | sort -n | tail -10

   5651 17th & Corcoran St NW
   5761 5th & K St NW
   6245 New Hampshire Ave & T St NW
   6997 Thomas Circle
   7267 14th & V St NW
   8092 15th & P St NW
   8975 Jefferson Dr & 14th St SW
   9419 Lincoln Memorial
  11183 Massachusetts Ave & Dupont Circle NW
  13880 Columbus Circle / Union Station


In [7]:
!csvcut -c5 2016q1.csv | csvstat

  1. Start station
	<class 'str'>
	Nulls: False
	Unique values: 370
	5 most frequent values:
		Columbus Circle / Union Station:	13120
		Massachusetts Ave & Dupont Circle NW:	9560
		Lincoln Memorial:	9388
		Jefferson Dr & 14th St SW:	8138
		Thomas Circle:	7479
	Max length: 64

Row count: 552399


In [8]:
!csvcut -c7 2016q1.csv | csvstat

  1. End station
	<class 'str'>
	Nulls: False
	Unique values: 370
	5 most frequent values:
		Columbus Circle / Union Station:	13880
		Massachusetts Ave & Dupont Circle NW:	11183
		Lincoln Memorial:	9419
		Jefferson Dr & 14th St SW:	8975
		15th & P St NW:	8092
	Max length: 64

Row count: 552399


### Part A: Answer

The 10 Capital Bikeshare stations that were the most popular departing stations in Q1 2016 were:
    1. Columbus Circle / Union Station  - 13120
    2. Massachusetts Ave & Dupont Circle NW - 9560
    3. Lincoln Memorial - 9388
    4. Jefferson Dr & 14th St SW - 8138
    5. Thomas Circle - 7479
    6. 15th & P St NW - 7401
    7. 14th & V St NW - 6568 
    8. New Hampshire Ave & T St NW - 6491
    9. Eastern Market Metro / Pennsylvania Ave & 7th St SE - 5649
    10. 17th & Corcoran St NW 5514
 
The 10 stations that were the most popular destination stations in Q1 2016 were:    
    1. Columbus Circle / Union Station  - 13880
    2. Massachusetts Ave & Dupont Circle NW - 11183
    3. Lincoln Memorial - 9419
    4. Jefferson Dr & 14th St SW - 8975
    5. 15th & P St NW - 8092
    6. 14th & V St NW - 7267
    7. Thomas Circle - 6997 
    8. New Hampshire Ave & T St NW - 6245
    9. 17th & Corcoran St NW - 5761
    10. 5th & K St NW - 5651

By utlizing the sort function I placed each field in the column next to its identical fields, then counted the values by filtering the fields to count the number of unique fields. In order to get the proper values outputted, I needed to sort the column by number not character and then get the last 10 because it sorts from least to greatest. I also used the statistical filter csvstat just to double check that my answers were correct for both the departing stations and destination stations because it gives the top 5 most frequent values. The majority of the most popular departing stations are also the most popular destination stations, this makes sense because they if you head to one of these places, chances are you are also going to depart from this station as well to leave. 

## Part B

In [8]:
!csvcut -c5,8 2016q1.csv | csvgrep -c1 -m "Columbus Circle" | sort | uniq -c | sort -n | tail -10 | csvlook -H

|------------------------------------------+----------|
|  column1                                 | column2  |
|------------------------------------------+----------|
|       15 Columbus Circle / Union Station | W21076   |
|       15 Columbus Circle / Union Station | W21450   |
|       15 Columbus Circle / Union Station | W22080   |
|       16 Columbus Circle / Union Station | W00714   |
|       16 Columbus Circle / Union Station | W20540   |
|       16 Columbus Circle / Union Station | W21239   |
|       16 Columbus Circle / Union Station | W21538   |
|       16 Columbus Circle / Union Station | W21641   |
|       16 Columbus Circle / Union Station | W21867   |
|       17 Columbus Circle / Union Station | W22227   |
|------------------------------------------+----------|


In [10]:
!csvcut -c7,8 2016q1.csv | csvgrep -c1 -m "Columbus Circle" | sort | uniq -c | sort -n | tail -10 | csvlook -H

|------------------------------------------+----------|
|  column1                                 | column2  |
|------------------------------------------+----------|
|       15 Columbus Circle / Union Station | W21867   |
|       15 Columbus Circle / Union Station | W21997   |
|       16 Columbus Circle / Union Station | W00714   |
|       16 Columbus Circle / Union Station | W20425   |
|       16 Columbus Circle / Union Station | W21076   |
|       16 Columbus Circle / Union Station | W21239   |
|       16 Columbus Circle / Union Station | W22080   |
|       16 Columbus Circle / Union Station | W22099   |
|       17 Columbus Circle / Union Station | W22227   |
|       18 Columbus Circle / Union Station | W00485   |
|------------------------------------------+----------|


### Part B: Answer

The 10 most popular bike for departing at Columbus Circle were:
    1. W22227
    2. W21867
    3. W21641
    4. W21538
    5. W21239
    6. W20540
    7. W00714
    8. W22080
    9. W21450
    10. W21076
The 10 most popular bike for arriving at Columbus Circle were:
    1. W00485
    2. W22227
    3. W22099
    4. W22080
    5. W21239
    6. W21076
    7. W20425
    8. W00714 
    9. W21997
    10. W21867

I grabbed the most popular departing station and destination station and only searched for bikes the corresponding bike number column respectively. Next I sorted by the columns we made in the cut ( and counted them. Again I took the tail so that I have the 10 most popular. I created default headers again to make the output more readable.

# Problem 3 - Filters

## Part A

In [4]:
!cat women.txt | python onewordperline.py | python lowercase.py | sort | uniq -c | sort -rn | head -25

   8155 and
   7689 the
   5152 to
   3523 of
   3245 her
   2774 it
   2503 in
   2447 you
   2343 she
   2233 for
   2033 was
   1978 as
   1937 that
   1854 with
   1598 he
   1469 but
   1362 jo
   1135 so
   1118 his
   1067 at
   1063 had
   1014 be
    976 on
    942 not
    916 if
sort: write failed: standard output: Broken pipe
sort: write error


Here we use two python files that split the words in separate lines and make all characters lowercase respectively. I used Idle, a python GUI, to write the formulas, I have used it in the past and it is what I am most comfortable using to write python code. We then use the same filters as before to count the total unique words and take the last 25 to find the stop words. There are 24 stop words that I will remove (the 25th word that was printed out was jo, which is not a stop word). The top 24 stop words I will remove next are:

    1. and
    2. the
    3. to
    4. of
    5. her
    6. it
    7. in
    8. you
    9. she
    10. for
    11. was
    12. as
    13. that
    14. with
    15. he
    16. but
    17. so
    18. his
    19. at
    20. had
    21. be
    22. on
    23. not
    24. if



## Part B

In [13]:
!cat women.txt| python onewordperline.py | python lowercase.py | python stopwords.py | sort | uniq -c | sort -rn | head -25

   1362 jo
    881 all
    843 my
    827 said
    821 is
    782 him
    755 me
    730 little
    725 one
    719 they
    717 have
    709 when
    708 do
    686 meg
    658 up
    652 amy
    632 by
    619 which
    604 we
    598 laurie
    591 like
    550 don
    544 no
    517 were
    513 or
sort: write failed: standard output: Broken pipe
sort: write error


Here we add an additional filter to remove the stop words that we found in Part A. We then print out the new top 25 words to ensure that the stop words we wanted to take out are all removed. 

For this problem I could not use "./file.py" and I am not sure why, but after researching I found that by adding python before the file name and then within the file taking out the ">>>" at the beginning and end, it read the files perfectly. It even gave me some syntax error help when it would not print out properly at first.